src/ApplicationInsights.Kubernetes/TelemetryInitializers/TelemetryKeyCache.cs

```diff
@@ -0,0 +1,34 @@
+using System.Collections.Concurrent;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.ApplicationInsights.Kubernetes
+{
+    internal class TelemetryKeyCache : ITelemetryKeyCache
+    {
+        private readonly AppInsightsForKubernetesOptions _options;
+
+        private ConcurrentDictionary<string, string> _telemetryKeyCache;
+
+        public TelemetryKeyCache(IOptions<AppInsightsForKubernetesOptions> options)
+        {
+            _options = options?.Value ?? throw new System.ArgumentNullException(nameof(options));
+
+            if (_options.TelemetryKeyProcessor != null)
+            {
+                _telemetryKeyCache = new ConcurrentDictionary<string, string>();
+            }
+        }
+
+        public string GetTelemetryProcessedKey(string originalKey)
+        {
+            // Implicit condition: telemetry cache stays null means the key processor is not set in options.
+            if (_telemetryKeyCache == null)
+            {
+                return originalKey;
+            }
+
+            return _telemetryKeyCache.GetOrAdd(originalKey, _options.TelemetryKeyProcessor.Invoke(originalKey));
```

Comment:
There is no limit on the key cache growth...

Reply:
Being lazy, I think I'll just put a cap on the dictionary.